keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ProbabilisticDoubleTensorContract.java

```diff
@@ -57,7 +69,7 @@ public static void sampleMethodMatchesLogProbMethod(Vertex<DoubleTensor> vertexU
             double percentage = (double) sampleBucket.getValue() / samples.length;
             double bucketCenter = sampleBucket.getKey();
 
-            double densityAtBucketCenter = Math.exp(vertexUnderTest.logProb(Nd4jDoubleTensor.scalar(bucketCenter)));
+            double densityAtBucketCenter = Math.exp(((Probabilistic)vertexUnderTest).logProb(Nd4jDoubleTensor.scalar(bucketCenter)));
```

Comment:
I know part of the goal of this PR was to remove the Probabilistic* abstract classes but you can avoid this awkward casting if you have a simple tagging class like:

```
public abstract class ProbabilisticDoubleVertex extends DoubleVertex implements ProbabilisticDouble {

    public ProbabilisticDoubleVertex(ValueUpdater<DoubleTensor> valueUpdater) {
        super(valueUpdater);
    }
}
```

and in Gaussian for example:

```
public class GaussianVertex extends ProbabilisticDoubleVertex {
```

This doesn't change anything functionally and you can still remove the code duplication but this would clean up the generics in a few places. Because the actual use cases for the methods on the DoubleVertex abstract class and the ProbabilisticDouble interface often coincide, I think this casting will happen quite a bit.

Reply:
Actually I can eliminate these casts by parametrising the method on `<V extends DoubleVertex & ProbabilisticDouble>`.

It's really important that we decouple the concepts of Probabilistic, Differentiable, Double/Integer/Boolean. Otherwise we get this combinatorial explosion of abstract classes that doesn't scale, leads to code duplication and increased complexity.